# Rollback-only

## 0. 前置说明

​    开发中偶遇：`Transaction rolled back because it has been marked as rollback-only` 按照正常的思路，抛出异常之后，`try-catch` 不抛出异常则事务正常提交。但是只时候出现了`rollback-only` 的异常。

## 1.情景复现

> **情景1**: 当本方法中抛出异常，会在本方法的事务中置 `Transaction` 的状态为只读状态，当一次Post请求结束的时候进行提交事务，会校验事务当前的状态，因抛出异常的时候已经置为只读状态，因此请求结束提交事务抛出Rollback-only 异常

```java
/**
 * 主入口
 */
@Transactional
public void test() throws BusinessException {
    for (int i = 0; i < 10; i++) {
        TransactionPo transactionPo = new TransactionPo();
        transactionPo.setId(String.valueOf(i));
        transactionPo.setName("test"+i);
        this.insertOne(transactionPo);
        try {
            // 另外一个Service 中跑出异常
            subTransaction.throwException();
        }catch (Exception e){
            e.printStackTrace();
        }
    }
}

/**
 * subTransaction 新开事务的注解，进行插入数据
 */
@Transactional
public void throwException(){
    System.out.println(1/0);
}
```


## 2. 问题解决
> 解决方法1：在调用的类中开启新的事务
```java
// 将subTransaction 中的throwException 事务改为：
@Transactional(propagation = Propagation.REQUIRES_NEW)
public void throwException(){
    System.out.println(1/0);
}
```

> 解决方法2：在当前类中 添加方法并且调用this.throwException 


## 3. 分析结果

> 1. rollback-only异常，是发生在异常本身才有可能出现，发生在子方法内部是不会出现的。因此这种现象最多是发生在事务嵌套里
> 2. 使用this 关键字使用事务，那么调用方法增加了事务，那么会沿用上一个事务
> 3. 切记：事务如果加在私有方法上是无法生效的

